datasource/ethereum/src/block_stream.rs

```diff
@@ -1096,5 +1225,188 @@ fn create_log_filter_from_subgraph(manifest: &SubgraphManifest) -> EthereumLogFi
+        })
+        .collect::<HashMap<H256, u64>>();
+
+    // Ensure all `Call` and `Log` triggers have a transaction index
+    for trigger in triggers.iter() {
+        match trigger {
+            EthereumTrigger::Log(log) => {
+                if !tx_hash_indexes.get(&log.transaction_hash.unwrap()).is_some() {
+                    return Err(format_err!("Unable to determine transaction index for Ethereum log."))
+                }
+            }
+            EthereumTrigger::Call(call) => {
+                if !tx_hash_indexes.get(&call.transaction_hash.unwrap()).is_some() {
+                    return Err(format_err!("Unable to determine transaction index for Ethereum call."))
+                }
+            }
+            EthereumTrigger::Block(_) => continue,
+        }
+    }
+
+    // Sort the triggers
+    triggers
+        .sort_by(|a, b| {
+            let a_tx_index = a.transaction_index(&tx_hash_indexes).unwrap();
+            let b_tx_index = b.transaction_index(&tx_hash_indexes).unwrap();
+            if a_tx_index.is_none() && b_tx_index.is_none() {
+                return Ordering::Equal
+            }
+            if a_tx_index.is_none() {
+                return Ordering::Greater
+            }
+            if b_tx_index.is_none() {
+                return Ordering::Less
+            }
+            a_tx_index.unwrap().cmp(&b_tx_index.unwrap())
+        });
+    
+    Ok(triggers)
+}
+
+fn parse_log_triggers(
+    log_filter: Option<EthereumLogFilter>,
+    block: &EthereumBlock
+) -> Vec<EthereumTrigger> {
+    if log_filter.is_none() {
+        return vec![]
+    }
+    let log_filter = log_filter.unwrap();
+    block
+        .transaction_receipts
+        .iter()
+        .flat_map(move |receipt| {
+            let log_filter = log_filter.clone();
+            receipt
+                .logs
+                .iter()
+                .filter(move |log| {
+                    log_filter.matches(log)
+                })
+                .map(move |log| {
+                    EthereumTrigger::Log(log.clone())
+                })
+        })
+        .collect()
+}
+
+fn parse_call_triggers(
+    call_filter: Option<EthereumCallFilter>,
+    block: &EthereumBlockWithCalls,
+) -> Vec<EthereumTrigger> {
+    if call_filter.is_none() || block.calls.is_none() {
+        return vec![]
+    }
+    let call_filter = call_filter.unwrap();
+    let calls = &block.calls.clone().unwrap();
+    calls
+        .iter()
+        .filter(move |call| {
+            call_filter.matches(call)
+        })
+        .map(move |call| {
+            EthereumTrigger::Call(call.clone())
+        })
+        .collect()
+}
+
+fn parse_block_triggers(
+    block_filter: Option<EthereumBlockFilter>,
+    block: &EthereumBlockWithCalls,
+) -> Vec<EthereumTrigger> {
+    if block_filter.is_none() {
+        return vec![]
+    }
+    let call_filter = EthereumCallFilter::from(block_filter.unwrap());
+    let calls = &block.calls.clone().unwrap();
```

Comment:
We're not checking whether `block.calls` is `None`, like we're doing above in `parse_call_triggers`. What we want to do here seems to be:
```rust
block_filter.map_or(vec![], |block_filter| {
    let call_filter = EthereumCallFilter::from(block_filter);
    block.calls.map_or(vec![], |calls| {
        calls
           .iter()
           ...
    })
})
```

Reply:
Updated to use more combinators